### PR TITLE
Rzil op root approach

### LINKED ISSUE #277: Redesign of ESIL, or use of alternatives
**Is your feature request related to a problem? Please describe.**
ESIL is imperfect (slow, no float support, no well tested).
In theory, if Rizin used a good IL, we could have a simplify analysis (as in ghidra).

"By modeling in this way, the analysis of different processors is put into a common framework, facilitating the development of retargetable analysis algorithms and applications" - ghidra

**Describe the solution you'd like**
Using alternative IL:
* p-code
* bil
* openreil
* ...

**Describe alternatives you've considered**
Modify the ESIL specification...

**Additional context**
https://ghidra.re/courses/languages/html/pcoderef.html

### PR DESCRIPTION
**Pull request by Heersin**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

1. [New IL](https://github.com/BinaryAnalysisPlatform/bap/pull/960) as an independent library which relies on `rz_util` only.
2. IL in core analysis are moved to `core/cil.c`.  (Related pr of moving esil -> #1360 )
3. Change the tech of collecting `stats` and `trace` in the old ESIL
4. Fix some bugs in the old pr #1361 #1451

**Test plan**

Should pass the old rizin tests.

**Closing issues**

closes #277 

**More Info**
0. Use tree-like approach
1. This pr created for `asan build` to detect bug.
2. Only enabled in `analysis_bf.c`. 
2. `Register arena` changes in other module (such as `debug`) are not covered in current new IL.

### CODE REVIEW

**Comment by XVilka** [comment]
Can you please use the branch from the repository instead? To trigger more CI jobs like AppVeyor, BSD, etc?

File: librz/analysis/rzil/rzil.c [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	return rzil;
+}
+
+/**
+ * Cleanup RZIL instance : clean VM, clean arch-specific user_data, and RZIL itself
+ * \param analysis pointer to rizin's RzAnalysis
+ * \param rzil pointer to RzAnalysisRzil
+ */
+RZ_API void rz_analysis_rzil_cleanup(RzAnalysis *analysis, RzAnalysisRzil *rzil) {
+	if (!rzil) {
```
Comment (XVilka) [inline librz/analysis/rzil/rzil.c:30]: Silent cleanup

File: librz/analysis/rzil/rzil.c [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+/**
+ * Set instruction pointer for the current RZIL VM session
+ * \param rzil RzAnalysis* pointer to RzAnalysisRzil instance
+ * \param addr ut64 address of new pc
+ * \return true if set successfully, else return false
+ */
+RZ_API bool rz_analysis_rzil_set_pc(RzAnalysisRzil *rzil, ut64 addr) {
+	if (rzil) {
```
Comment (XVilka) [inline librz/analysis/rzil/rzil.c:48]: Reverse the condition

File: librz/analysis/rzil/rzil.c [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+/**
+ * Set operand list starting at the specified address for the RZIL VM
+ * \param rzil RzAnalysisRzil* pointer to RzAnalysisRzil
+ * \param addr ut64 address of current pc
+ * \param oplist RzPvector* vector of core theory opcodes
+ */
+RZ_API void rz_analysis_set_rzil_op(RzAnalysisRzil *rzil, ut64 addr, RzPVector *oplist) {
+	if (!rzil) {
+		eprintf("uninitialized rzil, cannot set op\n");
```
Comment (XVilka) [inline librz/analysis/rzil/rzil.c:89]: `RZ_LOG_ERROR()`

File: librz/analysis/rzil/rzil_trace.c [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+			goto error;
+		}
+		if (b->bytes && a->bytes && b->size > 0) {
+			memcpy(b->bytes, a->bytes, b->size);
+		}
+		trace->arena[i] = b;
+	}
+	return trace;
+error:
+	eprintf("Fail to init RZIL trace\n");
```
Comment (XVilka) [inline librz/analysis/rzil/rzil_trace.c:62]: `RZ_LOG_ERROR()`

File: librz/core/cmd_analysis.c [hunk 5]
```diff
@@ -4836,6 +4836,79 @@ static void __analysis_esil_function(RzCore *core, ut64 addr) {
...
+	switch (input[0]) {
+	case 'r': // "aezr"
+		// 'aer' is an alias for 'ar'
+		// cmd_analysis_reg(core, input + 1);
+		eprintf("[WIP] Remove it or replace it ?");
+		break;
+	case 's': // "aezs"
+		switch (input[1]) {
+		case '?': // "aez?"
+			rz_core_cmd0(core, "ae?~aes");
```
Comment (XVilka) [inline librz/core/cmd_analysis.c:4871]: It should print the help, not like this.

File: librz/il/definitions/bag.c [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+/**
+ * Store an item into bag
+ * \param bag RzILBag, point to the bag
+ * \param item A pointer to an element you want to store
+ * \return true if store successfully
+ */
+RZ_API bool rz_il_add_to_bag(RZ_NONNULL RzILBag *bag, RZ_NONNULL void *item) {
+	rz_return_val_if_fail(bag && item, false);
+	if (bag->item_count >= bag->capcity) {
+		printf("[Cannot Carry More Values]\n");
```
Comment (XVilka) [inline librz/il/definitions/bag.c:88]: `RZ_LOG_ERROR()`

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by Heersin** [comment]
Sorry I pushed a temporary branch from my repo. I will create a real pr instead. Sorry about the noise

**Pull request closed without merging**